openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -189,27 +194,37 @@ def get_cohort(user, course_key, assign=True):
     Raises:
        ValueError if the CourseKey doesn't exist.
     """
+    request_cache = RequestCache.get_request_cache()
+    cache_key = u"user_{}_cohort".format(user.id)
+
+    if use_cached and cache_key in request_cache.data:
+        return request_cache.data[cache_key]
+
+    request_cache.data.pop(cache_key, None)
+
     # First check whether the course is cohorted (users shouldn't be in a cohort
     # in non-cohorted courses, but settings can change after course starts)
-    try:
-        course = courses.get_course_by_id(course_key)
-    except Http404:
-        raise ValueError("Invalid course_key")
-
-    if not course.is_cohorted:
-        return None
+    course_cohort_settings = get_course_cohort_settings(course_key)
+    if not course_cohort_settings.is_cohorted:
+        return request_cache.data.setdefault(cache_key, None)
 
+    # If course is cohorted, check if the user already has a cohort.
     try:
-        return CourseUserGroup.objects.get(
+        cohort = CourseUserGroup.objects.get(
             course_id=course_key,
             group_type=CourseUserGroup.COHORT,
             users__id=user.id,
         )
+        return request_cache.data.setdefault(cache_key, cohort)
     except CourseUserGroup.DoesNotExist:
-        # Didn't find the group.  We'll go on to create one if needed.
+        # Didn't find the group. If we do not want to assign, return here.
         if not assign:
+            # Do not cache the cohort here, because in the next call assign
+            # may be True, and we will have to assign the user a cohort.
             return None
 
+    # Otherwise assign the user a cohort.
+    course = courses.get_course(course_key)
```

Comment:
Isn't this really expensive? Do we need to pass the full course around, as opposed to the course key?